File: postman/sdks/Specs SDKs/Search Service/typescript/src/services/search-endpoints/models/releases-default-model.ts
import { z } from 'zod';
import {
  CurrentStatusModel,
  currentStatusModel,
  currentStatusModelRequest,
  currentStatusModelResponse,
} from './current-status-model';
import {
  ReleasesDefaultModelGenre,
  releasesDefaultModelGenre,
  releasesDefaultModelGenreRequest,
  releasesDefaultModelGenreResponse,
} from './releases-default-model-genre';
import {
  TrackOrReleaseLabelModel,
  trackOrReleaseLabelModel,
  trackOrReleaseLabelModelRequest,
  trackOrReleaseLabelModelResponse,
} from './track-or-release-label-model';
import {
  ReleaseTrackModel,
  releaseTrackModel,
  releaseTrackModelRequest,
  releaseTrackModelResponse,
} from './release-track-model';
import {
  ReleaseKeyModel,
  releaseKeyModel,
  releaseKeyModelRequest,
  releaseKeyModelResponse,
} from './release-key-model';
import {
  TrackOrReleaseArtistModel,
  trackOrReleaseArtistModel,
  trackOrReleaseArtistModelRequest,
  trackOrReleaseArtistModelResponse,
} from './track-or-release-artist-model';
import {
  ReleaseAggregatorModel,
  releaseAggregatorModel,
  releaseAggregatorModelRequest,
  releaseAggregatorModelResponse,
} from './release-aggregator-model';
import { PriceModel, priceModel, priceModelRequest, priceModelResponse } from './price-model';

/**
 * Zod schema for the ReleasesDefaultModel model.
 * Defines the structure and validation rules for this data type.
 * This is the shape used in application code - what developers interact with.
 */
export const releasesDefaultModel = z.lazy(() => {
  return z.object({
    score: z.number(),
    currentStatus: z.array(currentStatusModel).optional().nullable(),
    genre: releasesDefaultModelGenre.optional().nullable(),
    label: trackOrReleaseLabelModel,
    tracks: z.array(releaseTrackModel).optional().nullable(),
    key: z.array(releaseKeyModel).optional().nullable(),
    artists: z.array(trackOrReleaseArtistModel).optional().nullable(),
    aggregator: releaseAggregatorModel,
    availableWorldwide: z.number(),
    catalogNumber: z.string().optional().nullable(),
    createDate: z.string().optional().nullable(),
    encodedDate: z.string().optional().nullable(),
    exclusive: z.number(),
    exclusiveDate: z.string().optional().nullable(),
    streamingDate: z.string().optional().nullable(),
    preorderDate: z.string().optional().nullable(),
    labelManager: z.string().optional().nullable(),
    preOrderDate_1: z.string().optional().nullable(),
    publishDate: z.string(),
    releaseDate: z.string(),
    releaseId: z.number(),
    releaseName: z.string(),
    releaseType: z.string(),
    status: z.number(),
    upc: z.string().optional().nullable(),
    updateDate: z.string(),
    price: priceModel.optional(),
    isExplicit: z.boolean().optional().nullable(),
    trackCount: z.number().optional().nullable(),
    releaseImageUri: z.string().optional().nullable(),
    releaseImageDynamicUri: z.string().optional().nullable(),
    downloads: z.number().optional().nullable(),
    isHype: z.boolean().optional().nullable(),
    isPreOrder: z.boolean().optional().nullable(),
    plays: z.number().optional().nullable(),
  });
});

/**
 *
 * @typedef  {ReleasesDefaultModel} releasesDefaultModel
 * @property {number}
 * @property {CurrentStatusModel[]}
 * @property {ReleasesDefaultModelGenre}
 * @property {TrackOrReleaseLabelModel}
 * @property {ReleaseTrackModel[]}
 * @property {ReleaseKeyModel[]}
 * @property {TrackOrReleaseArtistModel[]}
 * @property {ReleaseAggregatorModel}
 * @property {number}
 * @property {string}
 * @property {string}
 * @property {string}
 * @property {number}
 * @property {string}
 * @property {string}
 * @property {string}
 * @property {string}
 * @property {string}
 * @property {string}
 * @property {string}
 * @property {number}
 * @property {string}
 * @property {string}
 * @property {number}
 * @property {string}
 * @property {string}
 * @property {PriceModel}
 * @property {boolean}
 * @property {number}
 * @property {string}
 * @property {string}
 * @property {number}
 * @property {boolean}
 * @property {boolean}
 * @property {number}
 */
export type ReleasesDefaultModel = z.infer<typeof releasesDefaultModel>;

/**
 * Zod schema for mapping API responses to the ReleasesDefaultModel application shape.
 * Handles any property name transformations from the API schema.
 * If property names match the API schema exactly, this is identical to the application shape.
 */
export const releasesDefaultModelResponse = z.lazy(() => {
  return z
    .object({
      score: z.number(),
      current_status: z.array(currentStatusModelResponse).optional().nullable(),
      genre: releasesDefaultModelGenreResponse.optional().nullable(),
      label: trackOrReleaseLabelModelResponse,
      tracks: z.array(releaseTrackModelResponse).optional().nullable(),
      key: z.array(releaseKeyModelResponse).optional().nullable(),
      artists: z.array(trackOrReleaseArtistModelResponse).optional().nullable(),
      aggregator: releaseAggregatorModelResponse,
      available_worldwide: z.number(),
      catalog_number: z.string().optional().nullable(),
      create_date: z.string().optional().nullable(),
      encoded_date: z.string().optional().nullable(),
      exclusive: z.number(),
      exclusive_date: z.string().optional().nullable(),
      streaming_date: z.string().optional().nullable(),
      preorder_date: z.string().optional().nullable(),
      label_manager: z.string().optional().nullable(),
      pre_order_date: z.string().optional().nullable(),
      publish_date: z.string(),
      release_date: z.string(),
      release_id: z.number(),
      release_name: z.string(),
      release_type: z.string(),
      status: z.number(),
      upc: z.string().optional().nullable(),
      update_date: z.string(),
      price: priceModelResponse.optional(),
      is_explicit: z.boolean().optional().nullable(),
      track_count: z.number().optional().nullable(),
      release_image_uri: z.string().optional().nullable(),
      release_image_dynamic_uri: z.string().optional().nullable(),
      downloads: z.number().optional().nullable(),
      is_hype: z.boolean().optional().nullable(),
      is_pre_order: z.boolean().optional().nullable(),
      plays: z.number().optional().nullable(),
    })
    .transform((data) => ({
      score: data['score'],
      currentStatus: data['current_status'],
      genre: data['genre'],
      label: data['label'],
      tracks: data['tracks'],
      key: data['key'],
      artists: data['artists'],
      aggregator: data['aggregator'],
      availableWorldwide: data['available_worldwide'],
      catalogNumber: data['catalog_number'],
      createDate: data['create_date'],
      encodedDate: data['encoded_date'],
      exclusive: data['exclusive'],
      exclusiveDate: data['exclusive_date'],
      streamingDate: data['streaming_date'],
      preorderDate: data['preorder_date'],
      labelManager: data['label_manager'],
      preOrderDate_1: data['pre_order_date'],
      publishDate: data['publish_date'],
      releaseDate: data['release_date'],
      releaseId: data['release_id'],
      releaseName: data['release_name'],
      releaseType: data['release_type'],
      status: data['status'],
      upc: data['upc'],
      updateDate: data['update_date'],
      price: data['price'],
      isExplicit: data['is_explicit'],
      trackCount: data['track_count'],
      releaseImageUri: data['release_image_uri'],
      releaseImageDynamicUri: data['release_image_dynamic_uri'],
      downloads: data['downloads'],
      isHype: data['is_hype'],
      isPreOrder: data['is_pre_order'],
      plays: data['plays'],
    }));
});

/**
 * Zod schema for mapping the ReleasesDefaultModel application shape to API requests.
 * Handles any property name transformations required by the API schema.
 * If property names match the API schema exactly, this is identical to the application shape.
 */
export const releasesDefaultModelRequest = z.lazy(() => {
  return z
    .object({
      score: z.number(),
      currentStatus: z.array(currentStatusModelRequest).optional().nullable(),
      genre: releasesDefaultModelGenreRequest.optional().nullable(),
      label: trackOrReleaseLabelModelRequest,
      tracks: z.array(releaseTrackModelRequest).optional().nullable(),
      key: z.array(releaseKeyModelRequest).optional().nullable(),
      artists: z.array(trackOrReleaseArtistModelRequest).optional().nullable(),
      aggregator: releaseAggregatorModelRequest,
      availableWorldwide: z.number(),
      catalogNumber: z.string().optional().nullable(),
      createDate: z.string().optional().nullable(),
      encodedDate: z.string().optional().nullable(),
      exclusive: z.number(),
      exclusiveDate: z.string().optional().nullable(),
      streamingDate: z.string().optional().nullable(),
      preorderDate: z.string().optional().nullable(),
      labelManager: z.string().optional().nullable(),
      preOrderDate_1: z.string().optional().nullable(),
      publishDate: z.string(),
      releaseDate: z.string(),
      releaseId: z.number(),
      releaseName: z.string(),
      releaseType: z.string(),
      status: z.number(),
      upc: z.string().optional().nullable(),
      updateDate: z.string(),
      price: priceModelRequest.optional(),
      isExplicit: z.boolean().optional().nullable(),
      trackCount: z.number().optional().nullable(),
      releaseImageUri: z.string().optional().nullable(),
      releaseImageDynamicUri: z.string().optional().nullable(),
      downloads: z.number().optional().nullable(),
      isHype: z.boolean().optional().nullable(),
      isPreOrder: z.boolean().optional().nullable(),
      plays: z.number().optional().nullable(),
    })
    .transform((data) => ({
      score: data['score'],
      current_status: data['currentStatus'],
      genre: data['genre'],
      label: data['label'],
      tracks: data['tracks'],
      key: data['key'],
      artists: data['artists'],
      aggregator: data['aggregator'],
      available_worldwide: data['availableWorldwide'],
      catalog_number: data['catalogNumber'],
      create_date: data['createDate'],
      encoded_date: data['encodedDate'],
      exclusive: data['exclusive'],
      exclusive_date: data['exclusiveDate'],
      streaming_date: data['streamingDate'],
      preorder_date: data['preorderDate'],
      label_manager: data['labelManager'],
      pre_order_date: data['preOrderDate_1'],
      publish_date: data['publishDate'],
      release_date: data['releaseDate'],
      release_id: data['releaseId'],
      release_name: data['releaseName'],
      release_type: data['releaseType'],
      status: data['status'],
      upc: data['upc'],
      update_date: data['updateDate'],
      price: data['price'],
      is_explicit: data['isExplicit'],
      track_count: data['trackCount'],
      release_image_uri: data['releaseImageUri'],
      release_image_dynamic_uri: data['releaseImageDynamicUri'],
      downloads: data['downloads'],
      is_hype: data['isHype'],
      is_pre_order: data['isPreOrder'],
      plays: data['plays'],
    }));
});

File: postman/sdks/Specs SDKs/Search Service/python/src/search_service_sdk/models/releases_default_model.py
from __future__ import annotations
from typing import Union
from typing import List
from pydantic import Field
from typing import Optional
from typing import Annotated
from pydantic import TypeAdapter
from .utils.base_model import BaseModel
from .current_status_model import CurrentStatusModel
from .track_or_release_label_model import TrackOrReleaseLabelModel
from .release_track_model import ReleaseTrackModel
from .release_key_model import ReleaseKeyModel
from .track_or_release_artist_model import TrackOrReleaseArtistModel
from .release_aggregator_model import ReleaseAggregatorModel
from .price_model import PriceModel
from .genre_model import GenreModel

# Union type for ReleasesDefaultModelGenre
# Pydantic will attempt to validate against each type in order
ReleasesDefaultModelGenre = Union[List[GenreModel], GenreModel]
# TypeAdapter for runtime validation of ReleasesDefaultModelGenre
# This allows validation of Union types which are not directly instantiable
from pydantic import TypeAdapter

ReleasesDefaultModelGenre_adapter = TypeAdapter(ReleasesDefaultModelGenre)


class ReleasesDefaultModel(BaseModel):
    """ReleasesDefaultModel

    :param score: score
    :type score: float
    :param current_status: current_status, defaults to None
    :type current_status: List[CurrentStatusModel], optional
    :param genre: genre, defaults to None
    :type genre: ReleasesDefaultModelGenre, optional
    :param label: label
    :type label: TrackOrReleaseLabelModel
    :param tracks: tracks, defaults to None
    :type tracks: List[ReleaseTrackModel], optional
    :param key: key, defaults to None
    :type key: List[ReleaseKeyModel], optional
    :param artists: artists, defaults to None
    :type artists: List[TrackOrReleaseArtistModel], optional
    :param aggregator: aggregator
    :type aggregator: ReleaseAggregatorModel
    :param available_worldwide: available_worldwide
    :type available_worldwide: int
    :param catalog_number: catalog_number, defaults to None
    :type catalog_number: str, optional
    :param create_date: create_date, defaults to None
    :type create_date: str, optional
    :param encoded_date: encoded_date, defaults to None
    :type encoded_date: str, optional
    :param exclusive: exclusive
    :type exclusive: int
    :param exclusive_date: exclusive_date, defaults to None
    :type exclusive_date: str, optional
    :param streaming_date: streaming_date, defaults to None
    :type streaming_date: str, optional
    :param preorder_date: preorder_date, defaults to None
    :type preorder_date: str, optional
    :param label_manager: label_manager, defaults to None
    :type label_manager: str, optional
    :param pre_order_date: pre_order_date, defaults to None
    :type pre_order_date: str, optional
    :param publish_date: publish_date
    :type publish_date: str
    :param release_date: release_date
    :type release_date: str
    :param release_id: release_id
    :type release_id: int
    :param release_name: release_name
    :type release_name: str
    :param release_type: release_type
    :type release_type: str
    :param status: status
    :type status: int
    :param upc: upc, defaults to None
    :type upc: str, optional
    :param update_date: update_date
    :type update_date: str
    :param price: price, defaults to None
    :type price: PriceModel, optional
    :param is_explicit: is_explicit, defaults to None
    :type is_explicit: bool, optional
    :param track_count: track_count, defaults to None
    :type track_count: int, optional
    :param release_image_uri: release_image_uri, defaults to None
    :type release_image_uri: str, optional
    :param release_image_dynamic_uri: release_image_dynamic_uri, defaults to None
    :type release_image_dynamic_uri: str, optional
    :param downloads: downloads, defaults to None
    :type downloads: int, optional
    :param is_hype: is_hype, defaults to None
    :type is_hype: bool, optional
    :param is_pre_order: is_pre_order, defaults to None
    :type is_pre_order: bool, optional
    :param plays: plays, defaults to None
    :type plays: int, optional
    """

    score: float
    current_status: Optional[List[CurrentStatusModel]] = Field(default=None)
    genre: Optional[ReleasesDefaultModelGenre] = Field(default=None)
    label: TrackOrReleaseLabelModel
    tracks: Optional[List[ReleaseTrackModel]] = Field(default=None)
    key: Optional[List[ReleaseKeyModel]] = Field(default=None)
    artists: Optional[List[TrackOrReleaseArtistModel]] = Field(default=None)
    aggregator: ReleaseAggregatorModel
    available_worldwide: int
    catalog_number: Optional[str] = Field(default=None)
    create_date: Optional[str] = Field(default=None)
    encoded_date: Optional[str] = Field(default=None)
    exclusive: int
    exclusive_date: Optional[str] = Field(default=None)
    streaming_date: Optional[str] = Field(default=None)
    preorder_date: Optional[str] = Field(default=None)
    label_manager: Optional[str] = Field(default=None)
    pre_order_date: Optional[str] = Field(default=None)
    publish_date: str
    release_date: str
    release_id: int
    release_name: str
    release_type: str
    status: int
    upc: Optional[str] = Field(default=None)
    update_date: str
    price: Optional[PriceModel] = Field(default=None)
    is_explicit: Optional[bool] = Field(default=None)
    track_count: Optional[int] = Field(default=None)
    release_image_uri: Optional[str] = Field(default=None)
    release_image_dynamic_uri: Optional[str] = Field(default=None)
    downloads: Optional[int] = Field(default=None)
    is_hype: Optional[bool] = Field(default=None)
    is_pre_order: Optional[bool] = Field(default=None)
    plays: Optional[int] = Field(default=None)
